<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Genotype by Environment analysis using statgenGxE • statgenGxE</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Genotype by Environment analysis using statgenGxE">
<meta property="og:description" content="statgenGxE">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">statgenGxE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li>
  <a href="../articles/index.html">Vignettes</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="../news/index.html">
    <span class="fa fa-newspaper-o"></span>
     
    news
  </a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-file-code-o"></span>
     
    functions
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-seedling"></span>
     
    other packages
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="https://biometris.github.io/statgenSTA">statgenSTA</a>
    </li>
    <li>
      <a href="https://biometris.github.io/statgenGWAS">statgenGWAS</a>
    </li>
  </ul>
</li>
<li>
  <a href="https://github.com/Biometris/statgenGxE">
    <span class="fa fa-github fa-lg"></span>
     
    github
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="statgenGxE_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Genotype by Environment analysis using statgenGxE</h1>
                        <h4 class="author">Bart-Jan van Rossum</h4>
            
            <h4 class="date">2020-09-25</h4>
      
      
      <div class="hidden name"><code>statgenGxE.Rmd</code></div>

    </div>

    
    
<div id="the-statgengxe-package" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#the-statgengxe-package" class="anchor"></a>The statgenGxE package</h1>
<p>The statgenGxE package is developed as an easy-to-use package for Genotype by Environment (GxE) analysis for data of plant breeding experiments with many options for plotting and reporting the results of the analyses.</p>
<p>This vignette describes how to perform the different types of analysis that are available in the package. The availability of functions in the package is based on the analyses described in <span class="citation">Malosetti, Ribaut, and van Eeuwijk (2013)</span>. Further suggested reading is <span class="citation">van Eeuwijk, Bustos-Korts, and Malosetti (2016)</span>.</p>
<p>The following types of analysis can be done using statgenGxE:</p>
<ul>
<li><a href="#vcomp">Mixed model analysis of GxE table of means</a></li>
<li><a href="#fw">Finlay-Wilkinson Analysis</a></li>
<li><a href="#am">AMMI Analysis</a></li>
<li><a href="#gge">GGE Analysis</a></li>
<li><a href="#me">Identifying mega environments</a></li>
<li><a href="#st">Stability measures</a></li>
<li><a href="#vcov">Modeling of heterogeneity of genetic variances and correlations</a></li>
</ul>
<p>For most of the analyses a pdf report can be created automatically, see <a href="#rep">Reporting</a>.</p>
<p><em>Note that due to technical restrictions the number of significant digits printed in tables throughout this vignette is not always optimal. In practice precision of the output can always be specified by the user.</em></p>
<hr>
</div>
<div id="dp" class="section level1">
<h1 class="hasAnchor">
<a href="#dp" class="anchor"></a>Data preparation</h1>
<p>The use of the package is demonstrated using maize data from the European Union project DROPS (<a href="https://cordis.europa.eu/project/id/244374" class="uri">https://cordis.europa.eu/project/id/244374</a>). The data is available from <a href="https://data.inra.fr/dataset.xhtml?persistentId=doi:10.15454/IASSTN&amp;version=2.0" class="uri">https://data.inra.fr/dataset.xhtml?persistentId=doi:10.15454/IASSTN&amp;version=2.0</a> <span class="citation">(Millet et al. 2019)</span> and the relevant data set is included as data.frame in the statgenGxE package.</p>
<p>The first step is loading the data into R.</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="kw">dropsPheno</span>)
</pre></div>
<p>dropsPheno contains data for the genotypic means (Best Linear Unbiased Estimators, BLUEs), with one value per genotype per experiment, for a selection of 10 experiments from the full data set and eight traits. These 10 experiments form a good representation of the full set of experiments covering the five scenarios described in <span class="citation">Millet et al. (2016)</span>. Throughout this vignette in all examples the trait grain.yield will be analyzed. For a more detailed description of the contents of the data see <code>help(dropsData)</code>.</p>
<p>The input for GxE analysis in the statgenGxE package is an object of class TD (Trial Data). For a detailed description on how to construct such an object see the vignette of the statgenSTA package (<code>vignette("Modeling field trials using statgenSTA")</code>). For GxE analysis it is enough to specify the genotype and trial option of the <code>createTD</code> function.</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="co">## Create a TD object from dropsPheno.</span>
<span class="kw">dropsTD</span> <span class="op">&lt;-</span> <span class="kw">statgenSTA</span>::<span class="fu"><a href="https://rdrr.io/pkg/statgenSTA/man/TD.html">createTD</a></span>(data = <span class="kw">dropsPheno</span>, genotype = <span class="st">"Variety_ID"</span>, trial = <span class="st">"Experiment"</span>)
</pre></div>
<p>Before doing any analysis, we can first have a look at the contents of the data. To explore heterogeneity of genetic variance we can create a box plot. Coloring the boxes by environmental scenario will provide valuable extra information.</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="co">## Create a box plot of dropsTD.</span>
<span class="co">## Color the boxes based on the variable scenarioFull.</span>
<span class="co">## Plot in  descending order.</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span>(<span class="kw">dropsTD</span>, plotType = <span class="st">"box"</span>, traits = <span class="st">"grain.yield"</span>, colorTrialBy = <span class="st">"scenarioFull"</span>,
     orderBy = <span class="st">"descending"</span>)
</pre></div>
<p><img src="statgenGxE_files/figure-html/TDbox-1.png" width="672"></p>
<p>From the plot it is clear that the trials in the hot, water deficient environments have a lower median and range than the other trials.</p>
<p>For further insight into the correlation structure between trials a scatter plot matrix can be made.</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="co">## Create a scatter plot of dropsTD.</span>
<span class="co">## Color the genotypes based on the variable geneticGroup.</span>
<span class="co">## Color the histograms for trials based on the variable scenarioFull.</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span>(<span class="kw">dropsTD</span>, plotType = <span class="st">"scatter"</span>, traits = <span class="st">"grain.yield"</span>, colorGenoBy = <span class="st">"geneticGroup"</span>, 
     colorTrialBy = <span class="st">"scenarioFull"</span>, 
     trialOrder = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Gai12W"</span>, <span class="st">"Kar13R"</span>, <span class="st">"Kar12W"</span>, <span class="st">"Kar13W"</span>, <span class="st">"Mar13R"</span>, <span class="st">"Mur13W"</span>,
                    <span class="st">"Mur13R"</span>, <span class="st">"Ner12R"</span>, <span class="st">"Cam12R"</span>, <span class="st">"Cra12R"</span>))
</pre></div>
<p><img src="statgenGxE_files/figure-html/TDscatter-1.png" width="816"></p>
<div id="coloring-of-genotypes-and-trials" class="section level2">
<h2 class="hasAnchor">
<a href="#coloring-of-genotypes-and-trials" class="anchor"></a>Coloring of genotypes and trials</h2>
<p>In all plots the default colors for both genotype groups and trial groups are chosen from a predefined color palette. For genotype groups the color palette is “Dark 2”, for trial groups it is “Alphabet”. See <a href="https://developer.r-project.org/Blog/public/2019/11/21/a-new-palette-for-r/index.html">here</a> for an overview of these colors.</p>
<p>It is possible to specify different colors for genotype groups and trial groups per plot using the options <code>colGeno</code> and <code>colTrial</code> respectively. Also, more conveniently, the default colors can be set using the options statgen.genoColors and statgen.trialColors.</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="co">## Set default colors for genotypes and trials.</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span>(<span class="st">"statgen.genoColors"</span> = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"blue"</span>, <span class="st">"green"</span>, <span class="st">"yellow"</span>))
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span>(<span class="st">"statgen.trialColors"</span> = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"red"</span>, <span class="st">"brown"</span>, <span class="st">"purple"</span>))
</pre></div>
<p>If a plot has more genotype groups than the number of colors specified as default colors, the default colors will be ignored and <code>topo.colors</code> will be used instead. For trial groups this is done in a similar way.</p>
<hr>
</div>
</div>
<div id="vcomp" class="section level1">
<h1 class="hasAnchor">
<a href="#vcomp" class="anchor"></a>Mixed model analysis of GxE table of means</h1>
<p>To investigate the structure of the genotype by environment data various mixed models can be fitted. In the statgenGxE package this can be done using the <code>gxeVarComp</code> function.</p>
<p>Six different types of models can be fitted depending on the structure of the environments in the data. These models are described in the table below, together with the function parameters used in <code>gxeVarComp</code> to fit the model.</p>
<table class="table">
<colgroup>
<col width="26%">
<col width="44%">
<col width="28%">
</colgroup>
<thead><tr class="header">
<th align="left">Structure of environments</th>
<th align="left">Model</th>
<th align="left">Function parameters</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Environments correspond to trials</td>
<td align="left">
<strong>trait</strong> = trial + <strong>genotype + genotype:trial</strong>
</td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left">Trials form a factorial structure of locations x years</td>
<td align="left">
<strong>trait</strong> = year + location + year:location + <strong>genotype + genotype:year + genotype:location + genotype:year:location</strong>
</td>
<td align="left"><code>locationYear = TRUE</code></td>
</tr>
<tr class="odd">
<td align="left">Trials are nested within year</td>
<td align="left">
<strong>trait</strong> = year + year:trial + <strong>genotype + genotype:year + genotype:year:trial</strong>
</td>
<td align="left"><code>nestingFactor = "year"</code></td>
</tr>
<tr class="even">
<td align="left">Trials are nested within locations</td>
<td align="left">
<strong>trait</strong> = location + location:trial + <strong>genotype + genotype:location + genotype:location:trial</strong>
</td>
<td align="left"><code>nestingFactor = "loc"</code></td>
</tr>
<tr class="odd">
<td align="left">Trials correspond to locations within regions across years</td>
<td align="left">
<strong>trait</strong> = region + region:location + year + region:year + region:location:year + <strong>genotype + genotype:region + genotype:region:location + genotype:year + genotype:region:year + genotype:region:location:year</strong>
</td>
<td align="left"><code>regionLocationYear = TRUE</code></td>
</tr>
<tr class="even">
<td align="left">Trials are nested within scenarios</td>
<td align="left">
<strong>trait</strong> = scenario + scenario:trial + <strong>genotype + genotype:scenario + genotype:scenario:trial</strong>
</td>
<td align="left"><code>nestingFactor = "scenario"</code></td>
</tr>
</tbody>
</table>
<p><em>For data in the form of GxE means, the last random term in all models above will become a residual term. If the GxE means are provided together with weights, then a residual term will be added to the models above. Be aware that when plot data are provided as input data, the mixed model analysis will be based on the assumption of completely randomized trials, which in almost all cases will not be appropriate for multi-environment trials.</em></p>
<p>All models can be fitted using either <code>lme4</code> or <code>asreml</code>. This can be specified using the <code>engine</code> parameter.</p>
<p>For diagnostic purposes and to identify the main sources of variation, all models are fitted three times:</p>
<ol style="list-style-type: decimal">
<li>With all terms fixed</li>
<li>With all terms random</li>
<li>As a mixed model, with the <strong>bold</strong> terms as random and the others as fixed</li>
</ol>
<div id="model-fitted-with-all-terms-as-fixed" class="section level3 unnumbered">
<h3 class="hasAnchor">
<a href="#model-fitted-with-all-terms-as-fixed" class="anchor"></a>Model fitted with all terms as fixed</h3>
<p>The function first fits a model where all model terms are included as fixed terms. Based on the ANOVA table of this model, terms in the fixed part of the model that are likely to give a problem when fitting the mixed model are removed because of the reduced connectivity and number of available observations to estimate that model term. Also a warning is printed if the mean sum of squares for a model term points to a possible zero variance component in the mixed model.</p>
</div>
<div id="model-fitted-with-all-terms-as-random" class="section level3 unnumbered">
<h3 class="hasAnchor">
<a href="#model-fitted-with-all-terms-as-random" class="anchor"></a>Model fitted with all terms as random</h3>
<p>Then a model is fitted where all model terms are included as random terms. Based on the variance components in this model the percentage of variance explained by each of the model components is determined. The percentages of variance are printed in the model summary, together with the variance components. The latter are presented on a standard deviation scale.</p>
</div>
<div id="mixed-model" class="section level3 unnumbered">
<h3 class="hasAnchor">
<a href="#mixed-model" class="anchor"></a>Mixed model</h3>
<p>Finally a mixed model is fitted as specified in the table above. Based on this model, variance components can be extracted, heritabilities on a line mean basis (across all trials) can be computed and predictions can be made. It is also possible to plot the results.</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="co">## Fit a model where trials are nested within scenarios.</span>
<span class="kw">dropsVarComp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gxeVarComp.html">gxeVarComp</a></span>(TD = <span class="kw">dropsTD</span>, trait = <span class="st">"grain.yield"</span>, nestingFactor = <span class="st">"scenarioFull"</span>)
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="kw">dropsVarComp</span>)
<span class="co">#&gt; Fitted model formula</span>
<span class="co">#&gt; grain.yield ~ scenarioFull + scenarioFull:trial + (1 | genotype) + (1 | genotype:scenarioFull) </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Sources of variation</span>
<span class="co">#&gt;                       component % variance expl.</span>
<span class="co">#&gt; scenarioFull               7.03          49.39 %</span>
<span class="co">#&gt; scenarioFull:trial         5.48          38.51 %</span>
<span class="co">#&gt; genotype                   0.81           5.71 %</span>
<span class="co">#&gt; genotype:scenarioFull      0.17           1.17 %</span>
<span class="co">#&gt; residuals                  0.74           5.23 %</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Analysis of Variance Table for fully fixed model</span>
<span class="co">#&gt;                         Df  Sum Sq Mean Sq   F value    Pr(&gt;F)    </span>
<span class="co">#&gt; scenarioFull             4 17405.4  4351.3 5799.7996 &lt; 2.2e-16 ***</span>
<span class="co">#&gt; scenarioFull:trial       5  6271.9  1254.4 1671.9308 &lt; 2.2e-16 ***</span>
<span class="co">#&gt; genotype               245  2320.6     9.5   12.6246 &lt; 2.2e-16 ***</span>
<span class="co">#&gt; genotype:scenarioFull  980  1034.0     1.1    1.4063 8.087e-09 ***</span>
<span class="co">#&gt; residuals             1225   919.1     0.8                        </span>
<span class="co">#&gt; ---</span>
<span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span>
</pre></div>
<p><em>Note that because the model is fitted with lme4, in this and further tables no standard errors are outputted. When using asreml for modeling standard errors will be available.</em></p>
<p>The diagnostics for the fitted model can be printed using the <code>diagnostics</code> function. This will print an incidence matrix of missing values for each of the terms in the random part of the fitted model.</p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="co">## Print diagnostics - output suppressed because of the large number of rows.</span>
<span class="fu"><a href="../reference/diagnostics.html">diagnostics</a></span>(<span class="kw">dropsVarComp</span>)
</pre></div>
<p>Extracting the variance components from the fitted model can be done using the <code>vc</code> function. The heritability (across trials) is computed using <code>herit</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="co">## Extract variance components.</span>
<span class="fu"><a href="../reference/vc.html">vc</a></span>(<span class="kw">dropsVarComp</span>)
<span class="co">#&gt;                       component</span>
<span class="co">#&gt; genotype                 0.8128</span>
<span class="co">#&gt; genotype:scenarioFull    0.1665</span>
<span class="co">#&gt; residuals                0.7444</span>
<span class="co">## Compute heritability.</span>
<span class="fu"><a href="../reference/herit.html">herit</a></span>(<span class="kw">dropsVarComp</span>)
<span class="co">#&gt; [1] 0.8169</span>
</pre></div>
<p>A plot can be made of the square roots of the variance component estimates. These are based on the fully random model.</p>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="co">## Plot the results of the fitted model.</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span>(<span class="kw">dropsVarComp</span>)
</pre></div>
<p><img src="statgenGxE_files/figure-html/VarCompPlot-1.png" width="672"> Predictions for the mixed model can be made for model terms that represent different levels of data aggregation; for the genotype main effect or for genotypic performance in grouped environments, or for genotypic performance in individual trials. These levels can be specified by the parameter <code>predictLevel</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="co">## Predictions of the genotype main effect.</span>
<span class="kw">predGeno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html">predict</a></span>(<span class="kw">dropsVarComp</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="kw">predGeno</span>)
<span class="co">#&gt;   genotype predictedValue</span>
<span class="co">#&gt; 1    11430          6.726</span>
<span class="co">#&gt; 2       A3          6.334</span>
<span class="co">#&gt; 3     A310          5.745</span>
<span class="co">#&gt; 4     A347          5.604</span>
<span class="co">#&gt; 5     A374          7.456</span>
<span class="co">#&gt; 6     A375          6.658</span>
<span class="co">## predictions at the level of genotype x scenarioFull.</span>
<span class="kw">predGenoTrial</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html">predict</a></span>(<span class="kw">dropsVarComp</span>, predictLevel = <span class="st">"scenarioFull"</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="kw">predGenoTrial</span>)
<span class="co">#&gt;   genotype scenarioFull predictedValue</span>
<span class="co">#&gt; 1    11430      WW.Cool         10.537</span>
<span class="co">#&gt; 2       A3      WW.Cool          9.701</span>
<span class="co">#&gt; 3     A310      WW.Cool          9.633</span>
<span class="co">#&gt; 4     A347      WW.Cool          8.999</span>
<span class="co">#&gt; 5     A374      WW.Cool         11.001</span>
<span class="co">#&gt; 6     A375      WW.Cool          9.981</span>
</pre></div>
<hr>
</div>
</div>
<div id="fw" class="section level1">
<h1 class="hasAnchor">
<a href="#fw" class="anchor"></a>Finlay-Wilkinson Analysis</h1>
<p>With the Finlay-Wilkinson Analysis <span class="citation">(Finlay and Wilkinson 1963)</span> we describe genotype by environment interaction by the heterogeneity of the slopes of a regression of individual genotypic performance on an environmental index. The environmental index is the average of all genotypes in an environment. The intercept expresses general performance across all environments, the slope represents adaptability, and the residuals may indicate a measure for stability.</p>
<p>The model fitted in the analysis is <span class="math inline">\(y_{ij} = \mu + G_i + \beta_iE_j + \epsilon_{ij}\)</span>, where <span class="math inline">\(y_{ij}\)</span> is the phenotypic value of genotype <span class="math inline">\(i\)</span> in environment <span class="math inline">\(j\)</span>, <span class="math inline">\(\mu\)</span> is the general mean, <span class="math inline">\(G_i\)</span> is the genotypic effect, <span class="math inline">\(\beta_i\)</span> a sensitivity parameters, <span class="math inline">\(E_j\)</span> the environment effect and <span class="math inline">\(\epsilon_{ij}\)</span> a residual.</p>
<p>In the statgenGxE package this analysis can be done using the <code>gxeFW</code> function. The model described above is fitted using an alternating regression algorithm. First, using starting values for <span class="math inline">\(\beta_i\)</span> and <span class="math inline">\(G_i\)</span>, <span class="math inline">\(E_j\)</span> is estimated. Next, <span class="math inline">\(E_j\)</span> is assumed known and <span class="math inline">\(\beta_i\)</span> and <span class="math inline">\(G_i\)</span> are estimated. This process is continued until convergence, i.e. until the change in <span class="math inline">\(\beta_i\)</span> between iterations is less then a specified tolerance (default 0.001). When estimating parameters, missing observations are estimated as well.</p>
<p>By default all trials in the <code>TD</code> object are used in the analysis, but this can be restricted using the parameter <code>trials</code>. The genotypes included in the analysis can be restricted using <code>genotypes</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit">
<span class="co">## Perform a Finlay-Wilkinson analysis for all trials.</span>
<span class="kw">dropsFW</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gxeFw.html">gxeFw</a></span>(TD = <span class="kw">dropsTD</span>, trait = <span class="st">"grain.yield"</span>)
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="kw">dropsFW</span>)
<span class="co">#&gt; Environmental effects </span>
<span class="co">#&gt; =====================</span>
<span class="co">#&gt;     trial      envEff se_envEff  envMean se_envMean rank</span>
<span class="co">#&gt; 1  Cam12R -4.93016532 0.0710838  1.97041   0.526746    9</span>
<span class="co">#&gt; 2  Cra12R -5.41700099 0.0710838  1.48358   0.564494   10</span>
<span class="co">#&gt; 3  Gai12W  4.29088420 0.0710838 11.19145   0.478808    1</span>
<span class="co">#&gt; 4  Kar12W  2.81602440 0.0710838  9.71659   0.378874    3</span>
<span class="co">#&gt; 5  Kar13R  2.98347060 0.0710838  9.88404   0.389212    2</span>
<span class="co">#&gt; 6  Kar13W  1.14169667 0.0710838  8.04227   0.298929    4</span>
<span class="co">#&gt; 7  Mar13R  0.83532848 0.0710838  7.73590   0.290526    5</span>
<span class="co">#&gt; 8  Mur13R -0.00198925 0.0710838  6.89858   0.280534    7</span>
<span class="co">#&gt; 9  Mur13W  0.53202700 0.0710838  7.43260   0.284630    6</span>
<span class="co">#&gt; 10 Ner12R -2.25027579 0.0710838  4.65030   0.346565    8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Anova </span>
<span class="co">#&gt; =====</span>
<span class="co">#&gt;                 Df Sum Sq Mean Sq  F value Pr(&gt;F)    </span>
<span class="co">#&gt; genotype       245   2321     9.5   11.986 &lt;2e-16 ***</span>
<span class="co">#&gt; trial            9  23677  2630.8 3329.261 &lt;2e-16 ***</span>
<span class="co">#&gt; Sensitivities  245    404     1.6    2.088 &lt;2e-16 ***</span>
<span class="co">#&gt; Residual      1960   1549     0.8                    </span>
<span class="co">#&gt; Total         2459  27951    11.4                    </span>
<span class="co">#&gt; ---</span>
<span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Most sensitive genotypes</span>
<span class="co">#&gt; ========================</span>
<span class="co">#&gt;   genotype    sens rank   se_sens genMean se_genMean MSdeviation</span>
<span class="co">#&gt;     Lo1251 1.34779    1 0.0904284 9.03024   0.280534    1.507974</span>
<span class="co">#&gt;   DK78371A 1.29340    2 0.0904284 7.34072   0.280534    0.313889</span>
<span class="co">#&gt;      PHG83 1.29337    3 0.0904284 7.79922   0.280534    1.447104</span>
<span class="co">#&gt;      FR697 1.27416    4 0.0904284 7.48625   0.280534    0.917458</span>
<span class="co">#&gt;  SC-Malawi 1.26473    5 0.0904284 7.12909   0.280534    1.272097</span>
</pre></div>
<p>Four types of plots can be made to investigate the output from the analysis. <code>plotType = "scatter"</code> creates three scatter plots where genotypic mean, square root of the mean squared deviation and sensitivity are plotted against each other.</p>
<div class="sourceCode" id="cb12"><pre class="downlit">
<span class="co">## Create scatter plot for Finlay Wilkinson analysis.</span>
<span class="co">## Color genotypes by geneticGroup.</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span>(<span class="kw">dropsFW</span>, plotType = <span class="st">"scatter"</span>, colorGenoBy = <span class="st">"geneticGroup"</span>)
</pre></div>
<p><img src="statgenGxE_files/figure-html/plotFWScatter-1.png" width="672"></p>
<p>With <code>plotType = "line"</code> a plot with fitted lines for all genotypes in the analysis is created.</p>
<div class="sourceCode" id="cb13"><pre class="downlit">
<span class="co">## Create line plot for Finlay Wilkinson analysis.</span>
<span class="co">## Color genotypes by geneticGroup.</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span>(<span class="kw">dropsFW</span>, plotType = <span class="st">"line"</span>, colorGenoBy = <span class="st">"geneticGroup"</span>)
</pre></div>
<p><img src="statgenGxE_files/figure-html/plotFWLine-1.png" width="672"></p>
<p><code>plotType = "trellis"</code> creates a trellis plot with observations and slopes per genotype. At most 64 genotypes are plotted. It is possible to select a subset of genotypes for plotting using the parameter <code>genotypes</code>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit">
<span class="co">## Create trellis plot for Finlay Wilkinson analysis.</span>
<span class="co">## Restrict to first 5 genotypes.</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span>(<span class="kw">dropsFW</span>, plotType = <span class="st">"trellis"</span>, genotypes = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"11430"</span>, <span class="st">"A3"</span>, <span class="st">"A310"</span>, <span class="st">"A347"</span>, <span class="st">"A374"</span>))
</pre></div>
<p><img src="statgenGxE_files/figure-html/plotFWTrellis-1.png" width="672"></p>
<p><code>plottType = "scatterFit"</code> creates a scatter plot of fitted values in the trial with the highest environmental effect against the fitted values in the trial with the lowest environmental effect.</p>
<div class="sourceCode" id="cb15"><pre class="downlit">
<span class="co">## Create scatter plot of fitted values for Finlay Wilkinson analysis.</span>
<span class="co">## Color genotypes by geneticGroup.</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span>(<span class="kw">dropsFW</span>, plotType = <span class="st">"scatterFit"</span>, colorGenoBy = <span class="st">"geneticGroup"</span>)
</pre></div>
<p><img src="statgenGxE_files/figure-html/plotFWScatterFit-1.png" width="672"></p>
<hr>
</div>
<div id="am" class="section level1">
<h1 class="hasAnchor">
<a href="#am" class="anchor"></a>AMMI Analysis</h1>
<p>The Additive Main Effects and Multiplicative Interaction (AMMI) model fits a model which involves the Additive Main effects (i.e. genotype and trial) along with Multiplicative Interaction effects. The additive effects are the classical ANOVA main effects for genotype and environment, the multiplicative effects follow from a principal component analysis on the interaction residuals (= genotype by environment means after adjustment for additive genotype and environment effects). This results in an interaction characterized by Interaction Principal Components (IPCA) enabling simultaneous plotting of genotypes and trials.</p>
<p>If the data contains missing values, those are imputed first using an iterative regression algorithm. This algorithm regresses each environment in turn on all others. This process is repeated until the difference between the fitted values in subsequent iterations becomes sufficiently small (see <code><a href="../reference/multMissing.html">help(multMissing)</a></code> for further details).</p>
<p>After imputation the model fitted in the analysis is <span class="math inline">\(y_{ij} = \mu + G_i + E_j + \Sigma_{m =1}^M\gamma_{mi}\delta_{mj} + \epsilon_{ij}\)</span>, where <span class="math inline">\(M\)</span> is the number of principal components, <span class="math inline">\(y_{ij}\)</span> is the phenotypic value of genotype <span class="math inline">\(i\)</span> in environment <span class="math inline">\(j\)</span>, <span class="math inline">\(\mu\)</span> is the general mean, <span class="math inline">\(G_i\)</span> is the genotypic effect, <span class="math inline">\(E_j\)</span> the environmental effect, <span class="math inline">\(\gamma_{mi}\)</span> the genotypic scores, <span class="math inline">\(\delta_{mj}\)</span> the environmental scores, and <span class="math inline">\(\epsilon_{ij}\)</span> a residual.</p>
<p>The AMMI analysis can be performed with the statgenGxE package using the function <code>gxeAmmi</code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit">
<span class="co">## Run gxeAmmi for grain.yield.</span>
<span class="kw">dropsAm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gxeAmmi.html">gxeAmmi</a></span>(TD = <span class="kw">dropsTD</span>, trait = <span class="st">"grain.yield"</span>)
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="kw">dropsAm</span>)
<span class="co">#&gt; Principal components </span>
<span class="co">#&gt; ====================</span>
<span class="co">#&gt;                            PC1     PC2</span>
<span class="co">#&gt; Standard deviation     1.59365 1.16125</span>
<span class="co">#&gt; Proportion of Variance 0.31860 0.16916</span>
<span class="co">#&gt; Cumulative Proportion  0.31860 0.48776</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Anova </span>
<span class="co">#&gt; =====</span>
<span class="co">#&gt; Analysis of Variance Table</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Response: grain.yield</span>
<span class="co">#&gt;                Df Sum Sq Mean Sq  F value Pr(&gt;F)    </span>
<span class="co">#&gt; Genotype      245   2321     9.5   10.694 &lt;2e-16 ***</span>
<span class="co">#&gt; Environment     9  23677  2630.8 2970.187 &lt;2e-16 ***</span>
<span class="co">#&gt; Interactions 2205   1953     0.9                    </span>
<span class="co">#&gt; PC1           253    622     2.5    4.182 &lt;2e-16 ***</span>
<span class="co">#&gt; PC2           251    330     1.3    2.238 &lt;2e-16 ***</span>
<span class="co">#&gt; Residuals    1701   1000     0.6                    </span>
<span class="co">#&gt; ---</span>
<span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Environment scores </span>
<span class="co">#&gt; ==================</span>
<span class="co">#&gt;               PC1        PC2</span>
<span class="co">#&gt; Cam12R  0.4422147  0.1214471</span>
<span class="co">#&gt; Cra12R  0.4357878  0.0228969</span>
<span class="co">#&gt; Gai12W  0.0693329 -0.4178891</span>
<span class="co">#&gt; Kar12W -0.0329114  0.2499027</span>
<span class="co">#&gt; Kar13R -0.3648773 -0.4132687</span>
<span class="co">#&gt; Kar13W -0.3620128 -0.3924541</span>
<span class="co">#&gt; Mar13R  0.0924811  0.1666174</span>
<span class="co">#&gt; Mur13R -0.2854567  0.3230107</span>
<span class="co">#&gt; Mur13W -0.3539324  0.5112075</span>
<span class="co">#&gt; Ner12R  0.3593742 -0.1714703</span>
</pre></div>
<p>With the default settings in the AMMI analysis a maximum of two principal components are used. This can be changed using the parameter <code>nPC</code> in the function. The number of principal components should be lower than the smallest of the number of genotypes or environments minus one. By specifying <code>nPC = NULL</code> the algorithm will determine the number of principal components by a method of forward selection.</p>
<div class="sourceCode" id="cb17"><pre class="downlit">
<span class="co">## Run gxeAmmi. Let algorithm determine number of principal components.</span>
<span class="kw">dropsAm2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gxeAmmi.html">gxeAmmi</a></span>(TD = <span class="kw">dropsTD</span>, trait = <span class="st">"grain.yield"</span>, nPC = <span class="kw">NULL</span>)
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="kw">dropsAm2</span>)
<span class="co">#&gt; Principal components </span>
<span class="co">#&gt; ====================</span>
<span class="co">#&gt;                            PC1     PC2     PC3      PC4      PC5      PC6      PC7</span>
<span class="co">#&gt; Standard deviation     1.59365 1.16125 1.03717 0.931089 0.825373 0.711735 0.661859</span>
<span class="co">#&gt; Proportion of Variance 0.31860 0.16916 0.13494 0.108750 0.085460 0.063550 0.054950</span>
<span class="co">#&gt; Cumulative Proportion  0.31860 0.48776 0.62270 0.731450 0.816910 0.880460 0.935410</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Anova </span>
<span class="co">#&gt; =====</span>
<span class="co">#&gt; Analysis of Variance Table</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Response: grain.yield</span>
<span class="co">#&gt;                Df Sum Sq Mean Sq  F value   Pr(&gt;F)    </span>
<span class="co">#&gt; Genotype      245   2321     9.5   10.694  &lt; 2e-16 ***</span>
<span class="co">#&gt; Environment     9  23677  2630.8 2970.187  &lt; 2e-16 ***</span>
<span class="co">#&gt; Interactions 2205   1953     0.9                      </span>
<span class="co">#&gt; PC1           253    622     2.5    9.280  &lt; 2e-16 ***</span>
<span class="co">#&gt; PC2           251    330     1.3    4.967  &lt; 2e-16 ***</span>
<span class="co">#&gt; PC3           249    264     1.1    3.994  &lt; 2e-16 ***</span>
<span class="co">#&gt; PC4           247    212     0.9    3.245  &lt; 2e-16 ***</span>
<span class="co">#&gt; PC5           245    167     0.7    2.570  &lt; 2e-16 ***</span>
<span class="co">#&gt; PC6           243    124     0.5    1.927 7.04e-10 ***</span>
<span class="co">#&gt; PC7           241    107     0.4    1.680 9.66e-07 ***</span>
<span class="co">#&gt; Residuals     476    126     0.3                      </span>
<span class="co">#&gt; ---</span>
<span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Environment scores </span>
<span class="co">#&gt; ==================</span>
<span class="co">#&gt;               PC1        PC2        PC3        PC4        PC5        PC6        PC7</span>
<span class="co">#&gt; Cam12R  0.4422147  0.1214471 -0.2369571  0.0930633 -0.2551905  0.1253164  0.1961217</span>
<span class="co">#&gt; Cra12R  0.4357878  0.0228969 -0.3241345  0.0408753 -0.3491380  0.1273883  0.0138441</span>
<span class="co">#&gt; Gai12W  0.0693329 -0.4178891 -0.1744131  0.2772191  0.6223073  0.3705224 -0.2762854</span>
<span class="co">#&gt; Kar12W -0.0329114  0.2499027  0.7045194  0.5170036 -0.1714308  0.1483767 -0.1121918</span>
<span class="co">#&gt; Kar13R -0.3648773 -0.4132687  0.0376676  0.0679819 -0.0676313 -0.0331790  0.7592167</span>
<span class="co">#&gt; Kar13W -0.3620128 -0.3924541 -0.0402238 -0.2244351 -0.5039275 -0.0615522 -0.5330438</span>
<span class="co">#&gt; Mar13R  0.0924811  0.1666174  0.3375440 -0.7634903  0.1925495  0.3287360  0.0748694</span>
<span class="co">#&gt; Mur13R -0.2854567  0.3230107 -0.1777747  0.0157987  0.1310997  0.0625228 -0.0122844</span>
<span class="co">#&gt; Mur13W -0.3539324  0.5112075 -0.3419304  0.0474776  0.1619466 -0.2876442 -0.0503901</span>
<span class="co">#&gt; Ner12R  0.3593742 -0.1714703  0.2157025 -0.0714941  0.2394149 -0.7804873 -0.0598564</span>
</pre></div>
<p>From the summary it is clear that the final number of principal components, determined by forward selection, is seven.</p>
<p>It is possible to exclude certain genotypes, e.g. outliers, from the analysis using the parameter <code>excludeGeno</code>.</p>
<div class="sourceCode" id="cb18"><pre class="downlit">
<span class="co">## Run gxeAmmi with three principal components.</span>
<span class="co">## Exclude genotypes 11430 and A3.</span>
<span class="kw">dropsAm3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gxeAmmi.html">gxeAmmi</a></span>(TD = <span class="kw">dropsTD</span>, trait = <span class="st">"grain.yield"</span>, nPC = <span class="fl">3</span>, 
                    excludeGeno = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"11430"</span>, <span class="st">"A3"</span>))
</pre></div>
<p>If the data contains a column year, it is possible to fit a separate AMMI model for each year. This can be done by specifying the parameter <code>byYear = TRUE</code>. This will run an AMMI analysis for all years in the data without taking into account the data for the other years.</p>
<div class="sourceCode" id="cb19"><pre class="downlit">
<span class="co">## Run gxeAmmi per year in the data.</span>
<span class="kw">dropsAmYear</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gxeAmmi.html">gxeAmmi</a></span>(TD = <span class="kw">dropsTD</span>, trait = <span class="st">"grain.yield"</span>, byYear = <span class="fl">TRUE</span>)
</pre></div>
<p>The results of the AMMI analysis can be displayed in two types of plot. “AMMI1” plots the first principal component against the main effects. “AMMI2” plots the second against the first principal component.</p>
<div class="sourceCode" id="cb20"><pre class="downlit">
<span class="co">## Create an AMMI1 plot.</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span>(<span class="kw">dropsAm</span>, plotType = <span class="st">"AMMI1"</span>)
</pre></div>
<p><img src="statgenGxE_files/figure-html/plotAmmi1-1.png" width="75%"></p>
<p>AMMI biplots appear in three forms <span class="citation">(Gower and Hand 1996)</span>:</p>
<ol style="list-style-type: decimal">
<li>genotype scaling, with emphasis on approximation of relations between genotypes (scale = 0)</li>
<li>environment scaling, with emphasis on approximation of relations between environments (scale = 1)</li>
<li>symmetric scaling, a compromise (scale = 0.5)</li>
</ol>
<div class="sourceCode" id="cb21"><pre class="downlit">
<span class="co">## Create an AMMI2 biplot with symmetric scaling.</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span>(<span class="kw">dropsAm</span>, scale = <span class="fl">0.5</span>, plotType = <span class="st">"AMMI2"</span>)
</pre></div>
<p><img src="statgenGxE_files/figure-html/plotAmmi2-1.png" width="75%"></p>
<p>The AMMI plot function has many options to customize the plot. It is possible to plot different principal components on the axes specifying <code>primAxis</code> and <code>secAxis</code> Genotypes can be grouped and colored by a variable in the data using <code>colorGenoBy</code>. Specify custom colors in <code>colGeno</code>. Similarly for coloring environments use <code>colorEnvBy</code> and <code>colEnv</code>.</p>
<div class="sourceCode" id="cb22"><pre class="downlit">
<span class="co">## Create an AMMI2 biplot.</span>
<span class="co">## Color genotypes based on variable geneticGroup. Use custom colors.</span>
<span class="co">## Color environments based on variable scenarioFull</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span>(<span class="kw">dropsAm</span>, scale = <span class="fl">0.4</span>, plotType = <span class="st">"AMMI2"</span>, 
     colorGenoBy = <span class="st">"geneticGroup"</span>, colGeno = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"red"</span>, <span class="st">"blue"</span>, <span class="st">"green"</span>, <span class="st">"yellow"</span>),
     colorEnvBy = <span class="st">"scenarioFull"</span>)
</pre></div>
<p><img src="statgenGxE_files/figure-html/plotAmmiCol-1.png" width="75%"></p>
<p>A convex hull can be plotted around the genotypes in an AMMI2 biplot.</p>
<div class="sourceCode" id="cb23"><pre class="downlit">
<span class="co">## Create an AMMI2 biplot with convex hull around the genotypes.</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span>(<span class="kw">dropsAm</span>, scale = <span class="fl">0.4</span>, plotType = <span class="st">"AMMI2"</span>, plotConvHull = <span class="fl">TRUE</span>, colorEnvBy = <span class="st">"scenarioFull"</span>)
</pre></div>
<p><img src="statgenGxE_files/figure-html/plotAmmiConvHull-1.png" width="75%"></p>
<p>Genotypes can be left out of the plot completely by setting <code>plotGeno = FALSE</code> and similarly <code>plotEnv = FALSE</code> assures no environments are plotted. For displaying genotypes by their names instead of points, use <code>sizeGeno</code> with a size larger than zero to indicate their text size. <code>envFactor</code> can be used to blow up the environmental scores in the plot. A value for <code>envFactor</code> between zero and one effectively blows up the genotypic scores. See <code><a href="../reference/plot.AMMI.html">help(plot.AMMI)</a></code> for full details on all plotting options.</p>
<hr>
</div>
<div id="gge" class="section level1">
<h1 class="hasAnchor">
<a href="#gge" class="anchor"></a>GGE Analysis</h1>
<p>A Genotype plus Genotype by Environment (GGE) analysis is very similar to an AMMI analysis. The difference is in the first step where, instead of genotype and environment, only environment is fitted as a main effect in the model. Therefore, the principal component analysis is performed on the genotype main effect and the interaction jointly. In essence, the GGE analysis fits a principal components model with two components to the two-way genotype by environment table with data centered per environment.</p>
<p>The model fitted in the GGE analysis is <span class="math inline">\(y_{ij} = \mu + E_j + \gamma_{1i}\delta_{1j} + \gamma_{2i}\delta_{2j} + \epsilon_{ij}\)</span>, where the parameters are as described in the AMMI analysis.</p>
<p>The GGE analysis is done in statgenGxE by running the function <code>gxeGGE</code> which has identical options as the <code>gxeAmmi</code> function.</p>
<div class="sourceCode" id="cb24"><pre class="downlit">
<span class="co">## Run gxeGGE with default settings.</span>
<span class="kw">dropsGGE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gxeGGE.html">gxeGGE</a></span>(TD = <span class="kw">dropsTD</span>, trait = <span class="st">"grain.yield"</span>)
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="kw">dropsGGE</span>) 
<span class="co">#&gt; Principal components </span>
<span class="co">#&gt; ====================</span>
<span class="co">#&gt;                            PC1     PC2</span>
<span class="co">#&gt; Standard deviation     3.29808 1.16854</span>
<span class="co">#&gt; Proportion of Variance 0.62358 0.07828</span>
<span class="co">#&gt; Cumulative Proportion  0.62358 0.70186</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Anova </span>
<span class="co">#&gt; =====</span>
<span class="co">#&gt; Analysis of Variance Table</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Response: grain.yield</span>
<span class="co">#&gt;                Df Sum Sq Mean Sq  F value Pr(&gt;F)    </span>
<span class="co">#&gt; Environment     9  23677  2630.8 1508.201 &lt;2e-16 ***</span>
<span class="co">#&gt; Interactions 2450   4274     1.7                    </span>
<span class="co">#&gt; PC1           253   2665    10.5   16.088 &lt;2e-16 ***</span>
<span class="co">#&gt; PC2           251    335     1.3    2.036 &lt;2e-16 ***</span>
<span class="co">#&gt; Residuals    1946   1274     0.7                    </span>
<span class="co">#&gt; ---</span>
<span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Environment scores </span>
<span class="co">#&gt; ==================</span>
<span class="co">#&gt;             PC1        PC2</span>
<span class="co">#&gt; Cam12R 0.129437  0.0379476</span>
<span class="co">#&gt; Cra12R 0.104238  0.0935894</span>
<span class="co">#&gt; Gai12W 0.278606  0.4476497</span>
<span class="co">#&gt; Kar12W 0.307360 -0.1771409</span>
<span class="co">#&gt; Kar13R 0.436810  0.3169631</span>
<span class="co">#&gt; Kar13W 0.429534  0.2866136</span>
<span class="co">#&gt; Mar13R 0.275734 -0.0465212</span>
<span class="co">#&gt; Mur13R 0.395051 -0.3667764</span>
<span class="co">#&gt; Mur13W 0.409046 -0.5838223</span>
<span class="co">#&gt; Ner12R 0.159916  0.3128596</span>
</pre></div>
<p>Options for plotting results of a GGE analysis are identical to those for an AMMI analysis. When a convex hull is plotted, lines from the origin perpendicular to the edges of the hull are added as well.</p>
<div class="sourceCode" id="cb25"><pre class="downlit">
<span class="co">## Create a GGE2 biplot.</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span>(<span class="kw">dropsGGE</span>, plotType = <span class="st">"GGE2"</span>)
</pre></div>
<p><img src="statgenGxE_files/figure-html/plotGGE-1.png" width="75%"></p>
<hr>
</div>
<div id="me" class="section level1">
<h1 class="hasAnchor">
<a href="#me" class="anchor"></a>Identifying mega environments</h1>
<p>For the computation of mega environments, an AMMI-2 model is fitted and then, using the fitted values from this model, the environments are clustered. Mega environments are created by grouping environments based on their best performing genotype. Environments that share the same best genotype belong to the same mega environment.</p>
<div class="sourceCode" id="cb26"><pre class="downlit">
<span class="co">## Compute mega environments.</span>
<span class="kw">dropsMegaEnv</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gxeMegaEnv.html">gxeMegaEnv</a></span>(TD = <span class="kw">dropsTD</span>, trait = <span class="st">"grain.yield"</span>)
<span class="co">## Summarize results.</span>
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="kw">dropsMegaEnv</span>)
<span class="co">#&gt; Mega environments based on grain.yield</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;  Mega factor  Trial Winning genotype AMMI estimates</span>
<span class="co">#&gt;    megaEnv 1 Kar12W           Lo1087        12.4621</span>
<span class="co">#&gt;    megaEnv 1 Mar13R           Lo1087        10.0328</span>
<span class="co">#&gt;    megaEnv 1 Mur13R           Lo1087        10.4911</span>
<span class="co">#&gt;    megaEnv 1 Mur13W           Lo1087        11.3054</span>
<span class="co">#&gt;    megaEnv 2 Kar13R           Lo1251        13.6593</span>
<span class="co">#&gt;    megaEnv 2 Kar13W           Lo1251        11.7879</span>
<span class="co">#&gt;    megaEnv 3 Gai12W           Lo1270        13.7960</span>
<span class="co">#&gt;    megaEnv 4 Cam12R            PHG35         4.2604</span>
<span class="co">#&gt;    megaEnv 4 Cra12R            PHG35         3.8492</span>
<span class="co">#&gt;    megaEnv 4 Ner12R            PHG35         7.0998</span>
</pre></div>
<p>As can be seen in the column Mega factor in the summary, four mega environments have been created, one for each of the winning genotypes “Lo1087”, “Lo1251”, “Lo1270” and “PHG35”.</p>
<p>The values for the Best Linear Unbiased Predictors (BLUPs) and associated standard errors for the genotypes based on the calculated mega environments, can be computed using the function <code>predict</code> on the output. This can be done using either lme4 <span class="citation">(Bates et al. 2015)</span> or asreml <span class="citation">(Butler et al. 2017)</span> as an engine for fitting the model.</p>
<div class="sourceCode" id="cb27"><pre class="downlit">
<span class="co">if</span> (<span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span>(package = <span class="st">"asreml"</span>, quietly = <span class="fl">TRUE</span>)) {
  <span class="co">## Compute BLUPs.</span>
  <span class="co">## Use asreml as engine for fitting model.</span>
  <span class="kw">geMegaEnvPred</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html">predict</a></span>(<span class="kw">dropsMegaEnv</span>, engine = <span class="st">"asreml"</span>)
  <span class="co">## Display BLUPs.</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="kw">geMegaEnvPred</span><span class="op">$</span><span class="kw">predictedValue</span>)
}
<span class="co">#&gt; License check Fri Sep 25 07:59:51 2020</span>
<span class="co">#&gt;       megaEnv 1 megaEnv 2 megaEnv 3 megaEnv 4</span>
<span class="co">#&gt; 11430    6.4742    7.2043    7.1343    6.7571</span>
<span class="co">#&gt; A3       5.9956    6.6085    6.7493    6.5720</span>
<span class="co">#&gt; A310     5.0941    4.3924    5.9140    6.8347</span>
<span class="co">#&gt; A347     5.1709    4.7004    5.6903    6.4463</span>
<span class="co">#&gt; A374     7.5322    7.3081    7.3062    7.2980</span>
<span class="co">#&gt; A375     6.3963    6.4869    6.8184    6.8900</span>
<span class="co">if</span> (<span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span>(package = <span class="st">"asreml"</span>, quietly = <span class="fl">TRUE</span>)) {
  <span class="co">## Display standard errors of the BLUPs.</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="kw">geMegaEnvPred</span><span class="op">$</span><span class="kw">standardError</span>)
}
<span class="co">#&gt;       megaEnv 1 megaEnv 2 megaEnv 3 megaEnv 4</span>
<span class="co">#&gt; 11430   0.33316   0.42862    0.3732   0.27448</span>
<span class="co">#&gt; A3      0.33316   0.42862    0.3732   0.27448</span>
<span class="co">#&gt; A310    0.33316   0.42862    0.3732   0.27448</span>
<span class="co">#&gt; A347    0.33316   0.42862    0.3732   0.27448</span>
<span class="co">#&gt; A374    0.33316   0.42862    0.3732   0.27448</span>
<span class="co">#&gt; A375    0.33316   0.42862    0.3732   0.27448</span>
</pre></div>
<p>The results of the predictions can be visualized in a scatter plot using the <code>plot</code> functions. This creates a similar plot to the one described in the <a href="#dp">data preparation</a> section earlier in the vignette. However, now the predicted values for the genotypes in the different mega environments are displayed. The predictions are calculated inside the plot functions. By default lme4 is used for this. To change to asreml specify <code>engine = asreml</code>. Genotypes can be colored by the value of a variable in the TD object used for calculating the mega environments by specifying <code>colorGenoBy</code>.</p>
<div class="sourceCode" id="cb28"><pre class="downlit">
<span class="co">if</span> (<span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span>(package = <span class="st">"asreml"</span>, quietly = <span class="fl">TRUE</span>)) {
  <span class="co">## Create a scatter plot of predictions in mega environments.</span>
  <span class="co">## Color genotypes based on geneticGroup.</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span>(<span class="kw">dropsMegaEnv</span>, engine = <span class="st">"asreml"</span>, colorGenoBy = <span class="st">"geneticGroup"</span>)
}
</pre></div>
<p><img src="statgenGxE_files/figure-html/scatterMegaEnv-1.png" width="672"></p>
<hr>
</div>
<div id="st" class="section level1">
<h1 class="hasAnchor">
<a href="#st" class="anchor"></a>Stability measures</h1>
<p>Different measures of stability can be calculated using the statgenGxE package, the cultivar-superiority measure of Lin &amp; Binns <span class="citation">(Lin and Binns 1988)</span>, Shukla’s stability variance <span class="citation">(Shukla 1972)</span> and Wricke’s ecovalence <span class="citation">(Wricke 1962)</span>.</p>
<p>The cultivar-superiority is a function of the sum of the squared differences between a cultivar’s mean and the best cultivar’s mean, where the sum is across trials.. Genotypes with the smallest values of the superiority tend to be more stable, and closer to the best genotype in each environment.</p>
<p>Shukla’s stability variance (static stability) is defined as the variance around the genotype’s phenotypic mean across all environments. This provides a measure of the consistency of the genotype, without accounting for performance.</p>
<p>Wricke’s Ecovalence Stability Coefficient is the contribution of each genotype to the GxE sum of squares, in an unweighted analysis of the GxE means. A low value indicates that the genotype responds in a consistent manner to changes in environment; i.e. is stable from a dynamic point of view.</p>
<p>To compute the stability measures use <code>gxeStability</code>. This will compute all three stability measures described above.</p>
<div class="sourceCode" id="cb29"><pre class="downlit">
<span class="co">## Compute stability measures for dropsTD.</span>
<span class="kw">dropsStab</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gxeStability.html">gxeStability</a></span>(TD = <span class="kw">dropsTD</span>, trait = <span class="st">"grain.yield"</span>)
<span class="co">## In the summary print the top two percent of the genotypes.</span>
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="kw">dropsStab</span>, pctGeno = <span class="fl">2</span>)
<span class="co">#&gt; </span>
<span class="co">#&gt; Cultivar-superiority measure (Top 2 % genotypes)</span>
<span class="co">#&gt;  genotype    mean superiority</span>
<span class="co">#&gt;      Pa36 4.03790     21.7865</span>
<span class="co">#&gt;      F252 4.35804     19.8752</span>
<span class="co">#&gt;     CO109 4.73382     17.6806</span>
<span class="co">#&gt;    EA1163 4.68129     17.4563</span>
<span class="co">#&gt;      EP72 4.83195     17.3885</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Static stability (Top 2 % genotypes)</span>
<span class="co">#&gt;   genotype    mean  static</span>
<span class="co">#&gt;     Lo1251 9.03024 20.7656</span>
<span class="co">#&gt;      PHG83 7.79922 19.1744</span>
<span class="co">#&gt;  SC-Malawi 7.12909 18.2355</span>
<span class="co">#&gt;      FR697 7.48625 18.1761</span>
<span class="co">#&gt;   DK78371A 7.34072 18.1680</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Wricke's ecovalence (Top 2 % genotypes)</span>
<span class="co">#&gt;  genotype    mean  wricke</span>
<span class="co">#&gt;      A310 5.64885 26.0203</span>
<span class="co">#&gt;    Lo1251 9.03024 24.1650</span>
<span class="co">#&gt;   UH_P064 6.81195 22.9318</span>
<span class="co">#&gt;    Lo1270 8.35135 22.2490</span>
<span class="co">#&gt;     PHG83 7.79922 20.0827</span>
</pre></div>
<p>Plotting the results yields a scatter plot for the square root of each stability measure, plotted against the genotypic mean.</p>
<div class="sourceCode" id="cb30"><pre class="downlit">
<span class="co">## Create plots for the different stability measures.</span>
<span class="co">## Color genotypes by geneticGroup.</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span>(<span class="kw">dropsStab</span>, colorGenoBy = <span class="st">"geneticGroup"</span>)
</pre></div>
<p><img src="statgenGxE_files/figure-html/plotStab-1.png" width="672"></p>
<hr>
</div>
<div id="vcov" class="section level1">
<h1 class="hasAnchor">
<a href="#vcov" class="anchor"></a>Modeling of heterogeneity of genetic variances and correlations</h1>
<p>The function <code>gxeVarCov</code> allows to fit linear mixed models of varying complexity to a set of trials without clear structure. When the trials can be organized by locations, years, regions or scenarios, then the mixed models in section <a href="#vcomp">2</a> are more appropriate. The objective of this function is to identify a model for the genetic variances and correlations across the trials that will help in quantifying genetic similarities between trials and in understanding the structure of genotype by environment interactions. The identified variance-covariance structure can be used as the random background in a multi-environment QTL model. Models can be fitted using either lme4 or asreml as modeling engine. If the engine is set to ‘lme4’, only a model considering multiple variance components, but with a compound symmetry variance-covariance structure can be fitted. If the engine is set to ‘asreml’, complex variance-covariance structures will be fitted sequentially to the genotype by environment data. The best fitting model is then identified.<br>
The models fitted are of the form <span class="math inline">\(y_{ij} = \mu_j + \epsilon_{ij}\)</span>, where <span class="math inline">\(y_{ij}\)</span> is the phenotypic value of genotype <span class="math inline">\(i\)</span> in environment <span class="math inline">\(j\)</span>, <span class="math inline">\(\mu_j\)</span> is the environmental mean, and <span class="math inline">\(\epsilon_{ij}\)</span> represents mainly genetic variation, although some non-genetic variation may be included as well. The random term <span class="math inline">\(\epsilon_{ij}\)</span> is modeled in eight ways as described in the table below (see <span class="citation">(Boer et al. 2007)</span>).</p>
<table class="table">
<colgroup>
<col width="15%">
<col width="26%">
<col width="17%">
<col width="17%">
<col width="23%">
</colgroup>
<thead><tr class="header">
<th align="left">Model</th>
<th align="left">Description</th>
<th align="left">var(<span class="math inline">\(g_{ij}\)</span>)</th>
<th align="left">cov(<span class="math inline">\(g_{ij}\)</span>;<span class="math inline">\(g_{ik}\)</span>)</th>
<th align="left">Number of parameters</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">identity</td>
<td align="left">identity</td>
<td align="left"><span class="math inline">\(\sigma_G^2\)</span></td>
<td align="left">0</td>
<td align="left">1</td>
</tr>
<tr class="even">
<td align="left">cs</td>
<td align="left">compound symmetry</td>
<td align="left"><span class="math inline">\(\sigma_G^2 + \sigma_{GE}^2\)</span></td>
<td align="left"><span class="math inline">\(\sigma_{GE}^2\)</span></td>
<td align="left">2</td>
</tr>
<tr class="odd">
<td align="left">diagonal</td>
<td align="left">diagonal matrix (heteroscedastic)</td>
<td align="left"><span class="math inline">\(\sigma_{GE_j}^2\)</span></td>
<td align="left">0</td>
<td align="left"><span class="math inline">\(J\)</span></td>
</tr>
<tr class="even">
<td align="left">hcs</td>
<td align="left">heterogeneous compound symmetry</td>
<td align="left"><span class="math inline">\(\sigma_G^2+\sigma_{GE_j}^2\)</span></td>
<td align="left"><span class="math inline">\(\sigma_G^2\)</span></td>
<td align="left"><span class="math inline">\(J+1\)</span></td>
</tr>
<tr class="odd">
<td align="left">outside</td>
<td align="left">heterogeneity outside</td>
<td align="left"><span class="math inline">\(\sigma_{G_j}^2\)</span></td>
<td align="left"><span class="math inline">\(\theta\)</span></td>
<td align="left"><span class="math inline">\(J+1\)</span></td>
</tr>
<tr class="even">
<td align="left">fa</td>
<td align="left">first order factor analytic</td>
<td align="left"><span class="math inline">\(\lambda_{1j}^2+\sigma_{GE_j}^2\)</span></td>
<td align="left"><span class="math inline">\(\lambda_{1j}\lambda_{1k}\)</span></td>
<td align="left"><span class="math inline">\(2J\)</span></td>
</tr>
<tr class="odd">
<td align="left">fa2</td>
<td align="left">second order factor analytic</td>
<td align="left"><span class="math inline">\(\lambda_{1j}^2+\lambda_{2j}^2+\sigma_{GE_j}^2\)</span></td>
<td align="left"><span class="math inline">\(\lambda_{1j}\lambda_{1k}+\lambda_{2j}\lambda_{2k}\)</span></td>
<td align="left"><span class="math inline">\(3J-1\)</span></td>
</tr>
<tr class="even">
<td align="left">unstructured</td>
<td align="left">unstructured</td>
<td align="left"><span class="math inline">\(\sigma_{G_j}^2\)</span></td>
<td align="left"><span class="math inline">\(\sigma_{G_{j,k}}^2\)</span></td>
<td align="left"><span class="math inline">\(J(J+1)/2\)</span></td>
</tr>
</tbody>
</table>
<p>In this table <span class="math inline">\(J\)</span> is the number of environments, <span class="math inline">\(\sigma_G^2\)</span> the variance component for the genotype main effects, <span class="math inline">\(\sigma_{GE}^2\)</span> the variance component for GxE interactions. <span class="math inline">\(\sigma_{G_j}^2\)</span> and <span class="math inline">\(\sigma_{GE_j}^2\)</span> are the environment specific variance components for the genotype main effects and GxE interaction in environment <span class="math inline">\(j\)</span>. <span class="math inline">\(\sigma_{G_{j,k}}^2\)</span> is the genetic covariance between environments <span class="math inline">\(j\)</span> and <span class="math inline">\(k\)</span>. <span class="math inline">\(\theta\)</span> is the common correlation between environments and <span class="math inline">\(\lambda_{1j}\)</span> and <span class="math inline">\(\lambda_{2j}\)</span> are environment specific multiplicative parameters.</p>
<p>The best model for the data is selected based on either the Akaike Information Criterion (AIC <span class="citation">(Akaike 1974)</span>) or the Baysian Information Criterion (BIC <span class="citation">(Schwarz 1978)</span>). Which criterion is used is determined by the parameter <code>criterion</code> in the function <code>gxeVarCov</code>.</p>
<div class="sourceCode" id="cb31"><pre class="downlit">
<span class="co">## Use lme4 for fitting the models - only compound symmetry.</span>
<span class="kw">dropsVC</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gxeVarCov.html">gxeVarCov</a></span>(TD = <span class="kw">dropsTD</span>, trait = <span class="st">"grain.yield"</span>)
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="kw">dropsVC</span>)
<span class="co">#&gt; Best model: cs, based on BIC.</span>
<span class="co">#&gt;         AIC     BIC Deviance NParameters</span>
<span class="co">#&gt; cs 7295.144 7306.76 7291.144           2</span>
</pre></div>
<div class="sourceCode" id="cb32"><pre class="downlit">
<span class="co">## Use asreml for fitting the models - eight models fitted. </span>
<span class="co">## Use AIC as criterion for determining the best model.</span>
<span class="co">if</span> (<span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span>(<span class="st">"asreml"</span>, quietly = <span class="fl">TRUE</span>)) {
  <span class="kw">dropsVC2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gxeVarCov.html">gxeVarCov</a></span>(TD = <span class="kw">dropsTD</span>, trait = <span class="st">"grain.yield"</span>, engine = <span class="st">"asreml"</span>, criterion = <span class="st">"AIC"</span>)
  <span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="kw">dropsVC2</span>)
}
<span class="co">#&gt; Best model: unstructured, based on AIC.</span>
<span class="co">#&gt;                   AIC      BIC Deviance NParameters</span>
<span class="co">#&gt; unstructured 2142.065 2455.706 2032.065          55</span>
<span class="co">#&gt; fa2          2296.191 2461.565 2238.191          29</span>
<span class="co">#&gt; fa           2409.822 2523.873 2369.822          20</span>
<span class="co">#&gt; outside      2415.683 2478.411 2393.683          11</span>
<span class="co">#&gt; hcs          2775.819 2838.548 2753.819          11</span>
<span class="co">#&gt; cs           2792.345 2803.750 2788.345           2</span>
<span class="co">#&gt; diagonal     3637.959 3694.985 3617.959          10</span>
<span class="co">#&gt; identity     3870.166 3875.869 3868.166           1</span>
</pre></div>
<p>As becomes clear from the summary, the best model based on AIC is the model with an unstructured variance-covariance structure. This indicates that GxE is present in the form of heterogeneity of genetic variances and correlations..</p>
<p><em>Note that for both factor analytic models to be fitted a minimum of five environments is needed. If the data contains less environments, those two models are skipped.</em></p>
<p>A heat map of the correlation between the environments based on the best fitted model can be plotted. It will be instructive to compare the fitted correlations of tis section with the ‘observed’ correlations in section <a href="#dp">1</a> and the GGE biplot in section <a href="#gge">5</a>. Comparisons with the mixed models in section <a href="#vcomp">2</a> are also relevant.</p>
<div class="sourceCode" id="cb33"><pre class="downlit">
<span class="co">if</span> (<span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span>(<span class="st">"asreml"</span>, quietly = <span class="fl">TRUE</span>)) {
  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span>(<span class="kw">dropsVC2</span>)
}
</pre></div>
<p><img src="statgenGxE_files/figure-html/geVCPlot-1.png" width="672"></p>
<p>In the plot correlations between environments are shown. A dark red color indicates a strong positive correlation between environments, a dark blue color a strong negative correlation. Environments are clustered by their correlations and ordered according to the results of the clustering.</p>
<hr>
</div>
<div id="fitted-values-and-residuals" class="section level1">
<h1 class="hasAnchor">
<a href="#fitted-values-and-residuals" class="anchor"></a>Fitted values and residuals</h1>
<p>Fitted values and residuals for the models fitted using AMMI, GGE, Finlay-Wilkinson, and the selected variance-covariance model can be extracted using the functions <code>fitted</code> and <code>residuals</code> respectively. These can be useful for investigating the quality of the fitted model by e.g. creating diagnostic plots of fitted values against residuals.</p>
<div class="sourceCode" id="cb34"><pre class="downlit">
<span class="co">## Extract the fitted values and residuals for the Finlay-Wilkinson model.</span>
<span class="kw">fitFW</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/fitted.values.html">fitted</a></span>(<span class="kw">dropsFW</span>)
<span class="kw">resFW</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/residuals.html">residuals</a></span>(<span class="kw">dropsFW</span>)

<span class="co">## Create a diagnostic plot of fitted values against residuals.</span>
<span class="kw">fitResFW</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html">merge</a></span>(<span class="kw">fitFW</span>, <span class="kw">resFW</span>, by = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"trial"</span>, <span class="st">"genotype"</span>))
<span class="kw">ggplot2</span>::<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="kw">fitResFW</span>, <span class="kw">ggplot2</span>::<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(x = <span class="kw">fittedValue</span>, y = <span class="kw">residual</span>,
                                       color = <span class="kw">trial</span>)) <span class="op">+</span>
  <span class="kw">ggplot2</span>::<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>()
</pre></div>
<p><img src="statgenGxE_files/figure-html/fitRes-1.png" width="672"></p>
<hr>
</div>
<div id="rep" class="section level1">
<h1 class="hasAnchor">
<a href="#rep" class="anchor"></a>Reporting</h1>
<p>For most of the analyses described in this vignette a pdf report can be made using the <code>report</code> function. Such a report typically consists of an overview of the analysis and parameters used and the most interesting results and plots for the analysis. Except for this pdf report also the .tex file used to generate this report and a folder containing the figures in the report, are generated. This will make it easy to adapt the report to the user’s specific needs.</p>
<p><em>Note that all reports are created using pdflatex. If no installation is found by <code><a href="https://rdrr.io/r/base/Sys.which.html">Sys.which("pdflatex")</a></code> the report function will generate an error.</em></p>
<p>For the analysis done in the vignette the reports below can be made. Specify a file with .pdf extension as <code>outfile</code> to set the location where the report is written.</p>
<div class="sourceCode" id="cb35"><pre class="downlit">
<span class="co">## Create a report for the Finlay Wilkinson analysis.</span>
<span class="fu"><a href="https://rdrr.io/pkg/statgenSTA/man/report.html">report</a></span>(<span class="kw">dropsFW</span>, outfile = <span class="st">"./myReports/FWReport.pdf"</span>)

<span class="co">## Create a report for the AMMI analysis.</span>
<span class="fu"><a href="https://rdrr.io/pkg/statgenSTA/man/report.html">report</a></span>(<span class="kw">dropsAm</span>, outfile = <span class="st">"./myReports/AMMIReport.pdf"</span>)

<span class="co">## Create a report for the GGE analysis.</span>
<span class="fu"><a href="https://rdrr.io/pkg/statgenSTA/man/report.html">report</a></span>(<span class="kw">dropsGGE</span>, outfile = <span class="st">"./myReports/GGEReport.pdf"</span>)

<span class="co">## Create a report for the stability analysis.</span>
<span class="fu"><a href="https://rdrr.io/pkg/statgenSTA/man/report.html">report</a></span>(<span class="kw">dropsStab</span>, outfile = <span class="st">"./myReports/stabReport.pdf"</span>)

<span class="co">## Create a report for the analysis of two-way GxE tables.</span>
<span class="fu"><a href="https://rdrr.io/pkg/statgenSTA/man/report.html">report</a></span>(<span class="kw">dropsVC2</span>, outfile = <span class="st">"./myReports/varCompReport.pdf"</span>)
</pre></div>
<hr>
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<div id="refs" class="references">
<div id="ref-Akaike1974">
<p>Akaike, H. 1974. “A New Look at the Statistical Model Identification.” <em>IEEE Transactions on Automatic Control</em> 19 (6): 716–23. <a href="https://doi.org/10.1109/tac.1974.1100705">https://doi.org/10.1109/tac.1974.1100705</a>.</p>
</div>
<div id="ref-Bates2015">
<p>Bates, Douglas, Martin Mächler, Ben Bolker, and Steve Walker. 2015. “Fitting Linear Mixed-Effects Models Using Lme4.” <em>Journal of Statistical Software</em> 67 (1). <a href="https://doi.org/10.18637/jss.v067.i01">https://doi.org/10.18637/jss.v067.i01</a>.</p>
</div>
<div id="ref-Boer2007">
<p>Boer, Martin P., Deanne Wright, Lizhi Feng, Dean W. Podlich, Lang Luo, Mark Cooper, and Fred A. van Eeuwijk. 2007. “A Mixed-Model Quantitative Trait Loci (QTL) Analysis for Multiple-Environment Trial Data Using Environmental Covariables for QTL-by-Environment Interactions, with an Example in Maize.” <em>Genetics</em> 177 (3): 1801–13. <a href="https://doi.org/10.1534/genetics.107.071068">https://doi.org/10.1534/genetics.107.071068</a>.</p>
</div>
<div id="ref-Gilmour2017">
<p>Butler, D. G., B. R. Cullis, A. R. Gilmour, B. G. Gogel, and R. Thompson. 2017. “ASReml-R Reference Manual Version 4.” <a href="https://www.vsni.co.uk/">https://www.vsni.co.uk/</a>.</p>
</div>
<div id="ref-Finlay1963">
<p>Finlay, KW, and GN Wilkinson. 1963. “The Analysis of Adaptation in a Plant-Breeding Programme.” <em>Australian Journal of Agricultural Research</em> 14 (6): 742. <a href="https://doi.org/10.1071/ar9630742">https://doi.org/10.1071/ar9630742</a>.</p>
</div>
<div id="ref-Gower1996">
<p>Gower, John, and David Hand. 1996. <em>Biplots</em>. London New York: Chapman &amp; Hall.</p>
</div>
<div id="ref-Lin1988">
<p>Lin, C. S., and M. R. Binns. 1988. “A Superiority Measure of Cultivar Performance for Cultivar Location Data.” <em>Canadian Journal of Plant Science</em> 68 (1): 193–98. <a href="https://doi.org/10.4141/cjps88-018">https://doi.org/10.4141/cjps88-018</a>.</p>
</div>
<div id="ref-Malosetti2013">
<p>Malosetti, Marcos, Jean-Marcel Ribaut, and Fred A. van Eeuwijk. 2013. “The Statistical Analysis of Multi-Environment Data: Modeling Genotype-by-Environment Interaction and Its Genetic Basis.” <em>Frontiers in Physiology</em> 4. <a href="https://doi.org/10.3389/fphys.2013.00044">https://doi.org/10.3389/fphys.2013.00044</a>.</p>
</div>
<div id="ref-Millet2019">
<p>Millet, Emilie J., Cyril Pommier, Mélanie Buy, Axel Nagel, Willem Kruijer, Therese Welz-Bolduan, Jeremy Lopez, et al. 2019. “A Multi-Site Experiment in a Network of European Fields for Assessing the Maize Yield Response to Environmental Scenarios.” Portail Data Inra. <a href="https://doi.org/10.15454/IASSTN">https://doi.org/10.15454/IASSTN</a>.</p>
</div>
<div id="ref-Millet2016">
<p>Millet, Emilie, Claude Welcker, Willem Kruijer, Sandra Negro, Stephane Nicolas, Sebastien Praud, Nicolas Ranc, et al. 2016. “Genome-Wide Analysis of Yield in Europe: Allelic Effects as Functions of Drought and Heat Scenarios.” <em>Plant Physiology</em>, July, pp.00621.2016. <a href="https://doi.org/10.1104/pp.16.00621">https://doi.org/10.1104/pp.16.00621</a>.</p>
</div>
<div id="ref-Schwarz1978">
<p>Schwarz, Gideon. 1978. “Estimating the Dimension of a Model.” <em>The Annals of Statistics</em> 6 (2): 461–64. <a href="https://doi.org/10.1214/aos/1176344136">https://doi.org/10.1214/aos/1176344136</a>.</p>
</div>
<div id="ref-Shukla1972">
<p>Shukla, G K. 1972. “Some Statistical Aspects of Partitioning Genotype-Environmental Components of Variability.” <em>Heredity</em> 29 (2): 237–45. <a href="https://doi.org/10.1038/hdy.1972.87">https://doi.org/10.1038/hdy.1972.87</a>.</p>
</div>
<div id="ref-vanEeuwijk2016">
<p>van Eeuwijk, Fred A., Daniela V. Bustos-Korts, and Marcos Malosetti. 2016. “What Should Students in Plant Breeding Know About the Statistical Aspects of Genotype Environment Interactions?” <em>Crop Science</em> 56 (5): 2119–40. <a href="https://doi.org/10.2135/cropsci2015.06.0375">https://doi.org/10.2135/cropsci2015.06.0375</a>.</p>
</div>
<div id="ref-Wricke1962">
<p>Wricke, G. 1962. “Evaluation Method for Recording Ecological Differences in Field Trials.” <em>Z Pflanzenzücht</em> 47: 92–96.</p>
</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Bart-Jan van Rossum.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
